libqtile/backend/x11/xcbq.py

```diff
@@ -759,9 +760,57 @@ def query_tree(self):
             parent = Window(self.conn, q.parent)
         return root, parent, [Window(self.conn, i) for i in q.children]
 
-    def paint_borders(self, color):
-        if color:
-            self.set_attribute(borderpixel=self.conn.color_pixel(color))
+    def paint_borders(self, colors, borderwidth, width, height):
+        if not colors or not borderwidth:
+            return
+
+        if isinstance(colors, str):
+            self.set_attribute(borderpixel=self.conn.color_pixel(colors))
+            return
+
+        if len(colors) > borderwidth:
+            colors = colors[:borderwidth]
+        core = self.conn.conn.core
+        outer_w = width + borderwidth * 2
+        outer_h = height + borderwidth * 2
+
+        with PixmapID(self.conn.conn) as pixmap:
+            with GContextID(self.conn.conn) as gc:
+                core.CreatePixmap(
+                    self.conn.default_screen.root_depth, pixmap, self.wid, outer_w, outer_h
+                )
+                core.CreateGC(gc, pixmap, 0, None)
+                borders = len(colors)
+                borderwidths = [borderwidth // borders] * borders
+                for i in range(borderwidth % borders):
+                    borderwidths[i] += 1
+                coord = 0
+                for i in range(len(borderwidths)):
+                    core.ChangeGC(
+                        gc, xcffib.xproto.GC.Foreground, [self.conn.color_pixel(colors[i])]
+                    )
+                    rect = xcffib.xproto.RECTANGLE.synthetic(
+                        coord, coord, outer_w - coord * 2, outer_h - coord * 2
+                    )
+                    core.PolyFillRectangle(pixmap, gc, 1, [rect])
+                    coord += borderwidths[i]
+                self.set_borderpixmap(pixmap, gc, borderwidth, width, height)
+
+    def set_borderpixmap(self, pixmap, gc, borderwidth, width, height):
+        core = self.conn.conn.core
+        outer_w = width + borderwidth * 2
+        outer_h = height + borderwidth * 2
+        with PixmapID(self.conn.conn) as border:
+            core.CreatePixmap(
+                self.conn.default_screen.root_depth, border, self.wid, outer_w, outer_h
+            )
+            most_w = outer_w - borderwidth
+            most_h = outer_h - borderwidth
+            core.CopyArea(pixmap, border, gc, borderwidth, borderwidth, 0, 0, most_w, most_h)
+            core.CopyArea(pixmap, border, gc, 0, 0, most_w, most_h, borderwidth, borderwidth)
+            core.CopyArea(pixmap, border, gc, borderwidth, 0, 0, most_h, most_w, borderwidth)
+            core.CopyArea(pixmap, border, gc, 0, borderwidth, most_w, 0, borderwidth, most_h)
+            core.ChangeWindowAttributes(self.wid, xcffib.xproto.CW.BorderPixmap, [border])
```

Comment:
X11 noob question here: why do we need two pixmaps and copy one into the other? Can't we just make the final one from the start?

Reply:
It is strange, I know, and is a weird quirk of painting onto X windows whose justification seems to have been lost to the ages. Basically, the pixmap maps onto the window assuming that the window's content (minus borders) is in the top left, the horizontal orders are both right at the bottom, and the vertical bottoms are at the far right. The 4 squares in the corners fit in there somewhere too. So we need to rearrange the contents of the pixmap before painting so that it maps how we want.